util/fs.go

```diff
@@ -0,0 +1,207 @@
+type extractedPackageFileSystem struct {
+	path string
+}
+
+func NewExtractedPackageFileSystem(p *Package) (*extractedPackageFileSystem, error) {
+	return &extractedPackageFileSystem{path: p.BasePath}, nil
+}
+
+func (fs *extractedPackageFileSystem) Stat(name string) (os.FileInfo, error) {
+	return os.Stat(filepath.Join(fs.path, name))
+}
+
+func (fs *extractedPackageFileSystem) Open(name string) (PackageFile, error) {
+	return os.Open(filepath.Join(fs.path, name))
+}
+
+func (fs *extractedPackageFileSystem) Glob(pattern string) (matches []string, err error) {
+	matches, err = filepath.Glob(filepath.Join(fs.path, pattern))
+	if err != nil {
+		return
+	}
+	for i := range matches {
+		matches[i] = matches[i][len(fs.path+string(filepath.Separator)):]
+	}
+	return
+}
+
+func (fs *extractedPackageFileSystem) Close() error { return nil }
+
+// zipPackageFileSystem provides utils to access files in a zipped package.
+type zipPackageFileSystem struct {
+	root   string
+	reader *zip.ReadCloser
+}
+
+func NewZipPackageFileSystem(p *Package) (*zipPackageFileSystem, error) {
+	reader, err := zip.OpenReader(p.BasePath)
+	if err != nil {
+		return nil, err
+	}
+	var root string
+	found := false
+	for _, f := range reader.File {
+		name := filepath.Clean(f.Name)
+		parts := strings.Split(name, string(filepath.Separator))
+		if len(parts) == 2 && parts[1] == "manifest.yml" {
+			root = parts[0]
+			found = true
+			break
+		}
+	}
+	if !found {
+		return nil, fmt.Errorf("failed to determine root directory in package (path: %s)", p.BasePath)
+	}
+	return &zipPackageFileSystem{
+		root:   root,
+		reader: reader,
+	}, nil
+}
+
+func (fs *zipPackageFileSystem) Stat(name string) (os.FileInfo, error) {
+	path := filepath.Join(fs.root, name)
+	f, err := fs.reader.Open(path)
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+	return f.Stat()
+}
+
+func (fs *zipPackageFileSystem) Open(name string) (PackageFile, error) {
+	path := filepath.Join(fs.root, name)
+	f, err := fs.reader.Open(path)
+	if err != nil {
+		return nil, err
+	}
+	return &zipFileSeeker{
+		File:   f,
+		reader: fs.reader,
+		path:   path,
+	}, nil
+}
+
+func (fs *zipPackageFileSystem) Glob(pattern string) (matches []string, err error) {
+	pattern = filepath.Join(fs.root, pattern)
+	for _, f := range fs.reader.File {
+		match, err := filepath.Match(pattern, filepath.Clean(f.Name))
+		if err != nil {
+			return nil, err
+		}
+		if match {
+			name := strings.TrimPrefix(f.Name, fs.root+string(filepath.Separator))
+			matches = append(matches, name)
+		}
+	}
+	return
+}
+
+func (fs *zipPackageFileSystem) Close() error {
+	return fs.reader.Close()
+}
+
+// zipFileSeeker implements the seeker interface for zip files.
+type zipFileSeeker struct {
+	fs.File
+
+	reader *zip.ReadCloser
+	path   string
+}
+
+// Seek implements the seeker interface for zip files. This is inefficient, it shouldn't
+// be frequently used.
+func (f *zipFileSeeker) Seek(offset int64, whence int) (n int64, err error) {
```

Comment:
Yes, this is hacky. Full story is that I wanted to use [`http.ServeContent`](https://pkg.go.dev/net/http#ServeContent) because it handles content type and so on, but this requires the file reader to implement `io.ReadSeeker`, an interface that zip file readers don't implement. So I added a minimal implementation for it.

I can give a thought to serve the file with the proper content type without `http.ServeContent`, what wouldn't require this hack, but would require to duplicat some things that this helper does.

Reply:
Ok, let's leave it for now. Thank you for the explanation.